src/Symfony/Component/Form/Console/Descriptor/Descriptor.php

```diff
@@ -0,0 +1,144 @@
+ */
+
+namespace Symfony\Component\Form\Console\Descriptor;
+
+use Symfony\Component\Console\Descriptor\DescriptorInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Form\ResolvedFormTypeInterface;
+use Symfony\Component\OptionsResolver\Exception\UndefinedOptionsException;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Yonel Ceruto <yonelceruto@gmail.com>
+ *
+ * @internal
+ */
+abstract class Descriptor implements DescriptorInterface
+{
+    /**
+     * @var SymfonyStyle
+     */
+    protected $output;
+    protected $type;
+    protected $ownOptions = array();
+    protected $overriddenOptions = array();
+    protected $inheritedOptions = array();
+    protected $extensionOptions = array();
+    protected $requiredOptions = array();
+    protected $parents = array();
+    protected $extensions = array();
+
+    /**
+     * {@inheritdoc}
+     */
+    public function describe(OutputInterface $output, $object, array $options = array())
+    {
+        $this->output = $output;
+
+        switch (true) {
+            case $object instanceof ResolvedFormTypeInterface:
+                $this->describeResolvedFormType($object, $options);
+                break;
+        }
+    }
+
+    abstract protected function describeResolvedFormType(ResolvedFormTypeInterface $resolvedFormType, array $options = array());
+
+    public function isRequiredOption($option)
+    {
+        return in_array($option, $this->requiredOptions, true);
+    }
+
+    public function hasParents()
+    {
+        return 0 !== count($this->parents);
+    }
+
+    public function getParents()
+    {
+        return array_keys($this->parents);
+    }
+
+    public function hasExtensions()
+    {
+        return 0 !== count($this->extensions);
+    }
+
+    public function getExtensions()
+    {
+        return array_keys($this->extensions);
+    }
+
+    public function collectOptions(ResolvedFormTypeInterface $type)
+    {
+        $this->parents = array();
+        $this->extensions = array();
+
+        $ownOptionsResolver = new OptionsResolver();
+        try {
+            $type->getInnerType()->configureOptions($ownOptionsResolver);
+        } catch (UndefinedOptionsException $e) {
+        }
+
+        if (null !== $type->getParent()) {
+            $optionsResolver = clone $this->getParentOptionResolver($type->getParent());
+        } else {
+            $optionsResolver = new OptionsResolver();
+        }
+
+        $this->ownOptions = array_diff($ownOptionsResolver->getDefinedOptions(), $optionsResolver->getDefinedOptions());
+        $this->overriddenOptions = array_intersect($ownOptionsResolver->getDefinedOptions(), $optionsResolver->getDefinedOptions());
+
+        $this->inheritedOptions = array();
+        foreach ($this->parents as $class => $opts) {
+            $this->inheritedOptions[$class] = array_diff($opts, $this->overriddenOptions);
```

Comment:
> I guess it's interesting knowing from which type originally came an option that was overridden.

You guess well :) it is missing right now, but at the end we need the diff to remove the overridden options from the inherited ones. Later, I'll do it :+1: